articles/huffman-coding-in-python/index.md

```diff
@@ -0,0 +1,184 @@
+
+The algorithm is based on the frequency with which each character in the file appears and the number of characters in a data structure with a frequency of 0. The Huffman encoding for a typical text file saves about 40%. We know that a file is stored on a computer as binary code and that each character in the file has been assigned a binary character, and the character codes usually have a fixed length for different characters. Huffman binary code, such as compiled executables, would therefore have a different space-saving. A binary file in which ASCII character is encoded with a frequency of 0.5 would have a very different distribution and frequency from its ASCII counterpart. 
+
+To compress a file with a sequence of characters, we need a table that gives us the sequences of bits used for each character. This table creates an encoding tree that uses the root/leaf path to create a bit sequence that encodes the characters. We can follow the roots and leaves to create a list of all characters with the maximum bit length of the encoded characters and the number of occurrences. 
+
+To construct an optimal tree, we use the greedy algorithm. Huffman encoding tree provides a minimum length of the character encodings for data compression. The nodes in the tree represent the frequency of a character's occurrence. The root node represents the length of the string, and traversing the tree gives us the character-specific encodings. Once the tree is constructed, traversing the tree gives us the respective codes for each symbol.
+
+The optimal tree upon completion is given in the image below:
+![optimal Tree](optimaltree.jpg)
+
+We will develop and implement a program that implements Huffman coding in the next section.  
+### Code
+
+We first define a class called `HuffmanCode` which is initialized with probabilities. Let us look at the flow of the code implemented below:
+
+1. Obtain the string and compute the frequency
+2. Using the frequency, obtain the probabilities
+3. Using the algorithm, compute the Huffman codes
+4. For the Huffman codes, compute mean length, variance, and entropy
+   
+Let us look at the algorithm used to compute the Huffman codes.
+
+1. We initially sort the probabilities in decreasing order. This is necessary to assign codes to each symbol according to their frequencies.
+2. Run a for loop from `0` to `length of string-2`
+3. For the first symbol assign `1` as the default code
+4. For the upcoming symbols, check for the previous symbols
+   1. If 1- add 0
+   2. If 11- add 1
+   3. If 10 - add 0
+   4. If 111 - add 1
+5. Once the Huffman codes are generated, read them in the reverse order to obtain the final_code. This is done to generate code as done by tree traversal.
+6. final_code is the list that contains all the huffman_codes in the order of the probabilities
+
+Let us look at the code now. There are a couple of functions defined inside the class which will be explained in the section given below.
```

Comment:
Let us look at the code now.  -> Let us now look at the code.

Makes the sentence sound less aggressive and threatening.